### PR TITLE
ssl: reuse same slice for SSL_read

### LINKED ISSUE #6617: Leak when reading from SSL sockets
**Envoy version**: v1.9.1

We're seeing an issue where Envoy sometimes leaks tens of gigabytes of memory. I still don't have a great understanding of why this is happening, but here's what we know so far.

Here's the output of pprof on a heap profile: https://gist.github.com/julia-stripe/04cee10b2185d81f1b84ad493920a9b3 which I think is saying that the memory is being allocated on [this line](https://github.com/envoyproxy/envoy/blame/53781eb42fdfce56ec3b187dc551344ab4f4a86f/source/extensions/transport_sockets/tls/ssl_socket.cc#L86).

Some other observations:

[Here are some logs from the heap profiler](https://gist.github.com/julia-stripe/37311c46088ab00a45143f6d3395292e). It looks like the pattern is that it allocates ~8GB of memory, frees some memory, and then does the same thing again. There are a lot of periods where it's allocating about 1GB/s.

tcmalloc doesn't seem to be doing a great job of releasing freed memory back to the OS though: this example of what's reported by the stats endpoint is pretty typical (16MB allocated, 17GB heap). 

```
$ curl localhost:9901/stats | grep memory
server.memory_allocated: 16697176
server.memory_heap_size: 17301504000
```

A few more relevant statistics that make this confusing:

* [output](https://gist.github.com/julia-stripe/176fcd4251af6237416d92b7e60a0ad6) of `curl localhost:9901/stats | grep bytes` (less than 100GB of data)
* [output](https://gist.github.com/julia-stripe/925b377e9dfc9d9b899ece3156fe5ec1) of `curl localhost:9901/stats | grep cx_total` (only ~2000 connections)

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

Description:
In some cases SSL_read will return a lot of small chunks, if we shrink the slice for every SSL_read, it will waste a lot of memory, reuse same slice until the slice is full in the same read event.

Risk Level: High
Testing: CI, tests provided in #6617
Docs Changes: N/A
Release Notes: N/A
Fixes #6617

### CODE REVIEW

**Comment by lizan** [comment]
Oops, intended to mark this as a draft, I'm still thinking how to test this...

**Comment by brian-pane** [comment]
@lizan thanks for the fix. This looks correct to me; I'm happy to go with either this or [PR6666](https://github.com/envoyproxy/envoy/pull/6666), whichever you prefer.

**Comment by lizan** [comment]
not stale, haven't get a chance to work on tests

**Comment by lizan** [comment]
@brian-pane @mattklein123 this is ready to review, added regression test.

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 1]
```diff
@@ -64,6 +64,29 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
...
 
+SslSocket::ReadResult SslSocket::sslReadIntoSlice(Buffer::RawSlice& slice) {
+  ReadResult result;
+  uint8_t* mem = static_cast<uint8_t*>(slice.mem_);
+  auto remaining = slice.len_;
+  while (remaining > 0) {
+    int rc = SSL_read(ssl_.get(), mem, remaining);
+    ENVOY_CONN_LOG(trace, "ssl read returns: {}", callbacks_->connection(), rc);
+    if (rc > 0) {
+      mem += rc;
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc]: Can you add `ASSERT(rc <= remaining);` ?

**Pull request merged**